training/train.py

```diff
@@ -304,113 +304,146 @@ def build_hyperparams(lower_limit, upper_limit, latent_factor,
-        for env in environments:
-            if env.get('name', '') == environment_name:
-                model_version = env.get('parameters', {}).get('MODEL_VERSION', '')
-                break
-    else:
-        _logger.error('ERROR - Invalid deployment prefix: "{}", supported values: {}'.format(
-            deployment_prefix, list(DEPLOYMENT_PREFIX_ENVIRONMENT_NAME_MAP.keys())))
 
-    _logger.info('Model version: {} for deployment prefix: {}'.format(
-        model_version, deployment_prefix))
-    return model_version
+def build_hyper_params_message(hyper_params):
+    """Build hyper params data string used for PR description and in yaml comments."""
+    return '- Hyper parameters :: {}'.format(json.dumps(hyper_params, indent=4, sort_keys=True))
 
 
-def create_git_pr(s3_client, hyper_params):  # pragma: no cover
+def format_body(body):
+    """Format PR body string to replace decorators."""
+    return body.replace('"', '').replace('{', '').replace('}', '').replace(',', '')
+
+
+def read_deployed_data(upstream_repo, s3_client, deployment_type):
+    """Read deployed data like yaml file, hyper params, model version."""
+    upstream_latest_commit_hash = upstream_repo.get_commits()[0].sha
+    _logger.info('Upstream latest commit hash: %s', upstream_latest_commit_hash)
+
+    contents = upstream_repo.get_contents(YAML_FILE_PATH, ref=upstream_latest_commit_hash)
+    yaml_dict = ruamel.yaml.load(contents.decoded_content.decode('utf8'),
+                                 ruamel.yaml.RoundTripLoader)
+
+    deployed_version = get_deployed_model_version(yaml_dict, deployment_type)
+    deployed_file_path = f'{deployed_version}/intermediate-model/hyperparameters.json'
+    deployed_hyperparams = s3_client.read_json_file(deployed_file_path)
+
+    deployed_data = {
+        'version': deployed_version,
+        'hyperparams': deployed_hyperparams
+    }
+    yaml_data = {
+        'commit_hash': upstream_latest_commit_hash,
+        'content_sha': contents.sha,
+        'dict': yaml_dict
+    }
+
+    return deployed_data, yaml_data
+
+
+def create_branch_and_update_yaml(deployment_type, deployed_data, yaml_data, hyper_params):
+    """Create branch and update yaml content on fork repo."""
+    # Update yaml model version for the given deployment
+    new_yaml_data = update_yaml_data(yaml_data['dict'], deployment_type,
+                                     MODEL_VERSION, hyper_params)
+    _logger.info('Modified yaml data, new length: %d', len(new_yaml_data))
+
+    # Connect to fabric8 analytic repo & get latest commit hash
+    f8a_repo = Github(GITHUB_TOKEN).get_repo(f'{FORK_REPO_NAME}/{PROJECT_NAME}')
+    _logger.info('f8a fork repo: %s', f8a_repo)
+
+    # Create a new branch on f8a repo
+    branch_name = f'bump_f8a-pypi-insights_for_{deployment_type}_to_{MODEL_VERSION}'
+    branch = f8a_repo.create_git_ref(f'refs/heads/{branch_name}', yaml_data['commit_hash'])
```

Comment:
Ideally the upstream repo reference should have been passed to this func and used the latest ref from that. This is making the function hard to read

Reply:
Doing this we read all upstream commits twice in the flow, first to read yaml and second at this place. Passing latest commit hash looked optimum.

Reply:
it hinders the readability, here the issue is not performance.